### ISSUE TITLE
Autoimport

### ISSUE THREAD

**Issue by bageljrkhanofemus** [davidhalter/jedi#1860]
I've been working on an autoimport plugin for [pylsp](https://github.com/python-lsp/python-lsp-server/pull/199). I'm wondering what would the feasibility of upstreaming this to mainline Jedi would be. The implementation consists of two parts:
1. The main library, in rope. Its written using python's ast. My new implementation is largely independent of rope's structure. This provides suggestions. 
2. The frontend, in pylsp. Its written using parso. This determines when to request for autoimport.

Jedi supporting autoimport would benefit a larger audience and merging should_import with jedi would deduplicate work.

**Comment by davidhalter** [comment]
That sounds great. If it's using parso already would it be possible to just push it into Jedi?

At the moment the problem is mostly that Jedi doesn't really work with a lot with caches, so it's hard to make it fast. So you would have to deal with that. I feel like if you work with parso anyways, that's a good start I guess.

I'm currently writing a Rust implementation of Jedi, but that will take a few more years until it's good. So any help here is appreciated! The code would probably belong in `jedi/api/refactoring/__init__.py`.

**Comment by bageljrkhanofemus** [comment]
1. The only two parts that use parso are for determing what names are already in a file and if a completion should be suggested. The majority of the code is part of the rope project, so either jedi would need to depend on rope for autoimport or copy and adapt the code for jedi. I would prefer if there wasn't two copies of the same code that I'm somewhat responsible for. 
2. Caching - it uses its own sqlite3 cache
3. Rust - that sounds like a cool project, would it be possible to contribute? Is it focused on LSP support or will it try to support as many editors as jedi does?
4. Location - it works by suggesting importtable names and should probably put in completion

**Comment by bageljrkhanofemus** [comment]
What alternative caching/indexing layer would you suggest instead of sqlite3?
Would it be possible to move autoimport to its own package so both rope and jedi can use it?
The current autoimport implementation doesn't require using the rest of rope if that's what you mean.

**Comment by davidhalter** [comment]
> Would it be possible to move autoimport to its own package so both rope and jedi can use it?

I'm not sure. Since Jedi already uses parso and manages its caches, I would strongly prefer this being part of Jedi.

> What alternative caching/indexing layer would you suggest instead of sqlite3?

I'm not sure what would fit Jedi to be honest... Maybe sqlite3 is the best there. I generally prefer simpler datastructures for such things (like a simple text file that you could search with grep with all of its definitions).

**Comment by davidhalter** [comment]
You are probably right about persistence. I'm also not sure if my approach would even be feasible. I have very good approaches for Rust, but there it's easier to structure memory.

**Comment by bageljrkhanofemus** [comment]
Its probably possible, but it'd be slower and harder for little gain.

**Comment by davidhalter** [comment]
Yeah, I guess sqlite3 is probably fine to have as an index, as long as invalidation works properly.

**Comment by bageljrkhanofemus** [comment]
I am working on that. I'll use the modified times of the packages to keep track of modules that need to be updated. I'm also going to release the rope autoimport code as a separate package and update this issue when that is done. Let me know if you have feedback, suggestions or any requirements that I should be aware of.

**Comment by davidhalter** [comment]
Just in case you're interested, Zuban has auto-import which works great there.
